packages/@aws-cdk/aws-apigateway/lib/authorizer.ts

```diff
@@ -1,9 +1,71 @@
+import { Construct, Lazy, Resource, ResourceProps } from '@aws-cdk/core';
+import { IRestApi } from '../lib/restapi';
+
+const CUSTOM_AUTH_SYMBOL = Symbol.for('@aws-cdk/apigateway.CustomAuthorizer');
+
 /**
  * Represents an API Gateway authorizer.
  */
 export interface IAuthorizer {
   /**
    * The authorizer ID.
+   * @attribute
    */
   readonly authorizerId: string;
+}
+
+/**
+ * Base class for all custom authorizers
+ */
+export abstract class CustomAuthorizer extends Resource implements IAuthorizer {
+
+  /**
+   * Check if the given object is of type CustomAuthorizer
+   */
+  public static isCustomAuthorizer(x: any): x is CustomAuthorizer {
+    return x !== null && typeof(x) === 'object' && CUSTOM_AUTH_SYMBOL in x;
+  }
+
+  /**
+   * The authorizer ID.
+   * @attribute
+   */
+  public abstract readonly authorizerId: string;
+
+  private _restApi?: IRestApi;
+
+  constructor(scope: Construct, id: string, props?: ResourceProps) {
+    super(scope, id, props);
+
+    Object.defineProperty(this, CUSTOM_AUTH_SYMBOL, { value: true });
+  }
+
+  /**
+   * Associate this authorizer to a specific {@link RestApi}.
+   * Typically, this is taken care of by the module while defining {@link Method}s via the {@link ResourceBase#addMethod | addMethod()} API call.
+   *
+   * @throws when a different RestApi is already associated to this authorizer.
+   */
+  public set restApi(restApi: IRestApi) {
```

Comment:
Why isn't this just a prop? If you want something like "attach", then see other examples in the module for how to implement this as an internal method. A public setter like this is an anti pattern.

Reply:
I've taken a look at the other patterns in the module, but the challenge here is --

`defaultMethodOptions.authorizer` is an attribute in the `RestApi` c'tor, while the authorizer itself needs the `restApiId` to generate a valid CF resource. 

So to resolve this deadlock, the authorizer props does not take the restApi (or restApiId) during construction and uses a `Lazy`. This means an instance of `TokenAuthorizer` can be passed in as part of `RestApi` construction. Calling `addMethod()` on any `Resource` of the `RestApi` would then pass the `restApiId` to the authorizer, thereby allowing the `Lazy` to be resolved.

If there's another pattern doing the same thing, please point me that way.

Reply:
I believe a private "bind" or "attach" method would work here. It's basically the same as what you did here but instead of a public setter it's an internal method called by the `RestApi` when the authorizer is "attached" to it:

```ts
/** @internal */
public _bind(api: IRestApi) {
  // .. do your thing
}
```